### ISSUE TITLE
[Mad Scientist Idea] Convert Engine to Cython

### ISSUE THREAD

**Issue by zicklag** [MinecoreDevTeam/Engine#4]
## Preface
![mad-scientist svg](https://user-images.githubusercontent.com/25393315/35781818-a17d125e-09b4-11e8-947a-e3551bb44f42.png)

OK, before I go any further I want to note that this may be a crazy and/or unfeasible idea, but I believe that, in light of the potential benefits, it is worth bringing up. Also, it would be good for you to know that, though I am an experienced programmer ( Python being my primary language, with some experience in Javascript, Lua, PHP, and Java), my experience with c++ is *very* minimal; I do not currently have the knowledge to make a correctly educated guess on what the true scope of what I am about to suggest is.

I am not saying that this is necessarily a good idea. I'm not saying that this is actually feasible to attempt on this project. This is just a "Mad Scientist Idea" that I think is just possible enough and has enough benefits that it would be great to consider. I would love to get some feedback from somebody who maybe knows more about c++ and Minetest core in particular.

## The Idea

What I am proposing is either a partial or complete transformation of the MineCore Engine source code from c++ to [Cython](http://cython.org/). From the [Cython documentation](https://cython.readthedocs.io/en/latest/src/quickstart/overview.html):

> [Cython] is a programming language that makes writing C extensions for the Python language as easy as Python itself. It aims to become a superset of the [Python] language which gives it high-level, object-oriented, functional, and dynamic programming. Its main feature on top of these is support for optional static type declarations as part of the language. The source code gets translated into optimized C/C++ code and compiled as Python extension modules. This allows for both very fast program execution and tight integration with external C libraries, while keeping up the high programmer productivity for which the Python language is well known.

This is obviously a potentially very large undertaking. It could mean changing the fundamental structure of a lot of the Engine code. Honestly I am not 100% sure what it would involve because I am not yet well acquainted with the current codebase, c++, or cmake and the build system. Because this would be such a large change, I should obviously provide some *very* good reasons for making the conversion. This is exactly what I hope to do.

### Cython is *Easier* than c++

Cython is easier and faster to write than c++. That obviously depends on who you are and whether or not you know Python to begin with, but Python ( which Cython is a super-set of ) is a high level language and c++ is a low level language. This makes Cython leaps and bounds easier to learn and write than c++. In the context of MineCore this has an even greater benefit than it would in a project that already has a strong community.

In a community like MineCore or Minetest, I believe one of the biggest obstacles to the project has been the lack of devs that can contribute to the engine by writing c++. You have *tons* of people who can write the Lua mods for Minetest, but a much smaller group of people that can actually improve the engine. When using the basic Python subset of Cython, Cython is just as easy to learn as the Lua language is. By making the engine programming language more approachable to a larger number of people you provide for a larger and more involved community.

The benefit of ease is not just useful for new or inexperienced developers. Even for experienced developers, Cython allows you to write features faster with less code. This allows for increase innovation and for spending more time being creative.

### Cython is just as powerful as c++

Unlike Python, Cython gets compiled to c/c++ code, which is then compiled by a c/c++ compiler. This has a number of benefits.

#### Cython can wrap native c/c++

You can use any c/c++ library in Cython. This means that the c++ libraries that are currently being used, such as Irrlicht, could still be used. Portions of the game could even be left written in c or c++ without any problem.

#### Cython is as fast as c++

By using static c++ type declarations, Cython code can run nearly as fast as c++ code. Cython also comes with a tool that does annotations to show you where you code is running slower than it could so you can intelligently decide where to put the static type declarations. This allows you to balance readability and simplicity with performance. If it becomes necessary  for performance you can still write pure c++ code and seamlessly incorporate it with the rest of the Cython codebase.

### There is the potential for Cython mods

I am not 100% sure about this, but I believe that it is possible to setup Cython mods. This would mean that you could write high performance mods for MineCore in a language that is not much more difficult than Lua. In addition, you could potentially utilize the entire core API to make mods instead of being limited to the provided Lua API. Cython mods would be almost as powerful as the engine itself.

You would not need to replace Lua modding but instead supplement it and allow for the option of higher performance and more powerful mods while maintaining the existing Lua modding community.

### Cython can be incrementally adopted

Because Cython gets compiled to c++, you don't need to convert the entire engine at the same time. You could incrementally convert different pieces of the engine to Cython. The Cython files would get compiled to c++ before the engine is compiled using CMake.

## Summary

I believe that there are a lot of good reasons to adopt Cython in MineCore. Honestly I don't see any disadvantage to Cython over c++. That being said I don't know that it is a feasible idea for MineCore because of the work that might be involved. Nevertheless, I think that it could help with one of the greatest needs of MineCore, community. Opening up core to a larger community by making its programming language easier could have a big impact on the projects ability to grow.

I know that this could be a major undertaking. I plan on doing a proof of concept in my own fork of the project to see if I can get an idea of what it would take to implement Cython in the project. If anybody who knows c++ is interested or even has some big reasons not to do this I would love to hear from them.

## Postscript

I have heard that the MineCoreDevTeam was looking for programmers to join the project. I can't guarantee that I have a lot of time to spend working on this, but I am greatly interested in this project. Me, @tekhnaeraav, and @harsulinsghost are very much aligned with the goals of the project and would like to help in any way that we can find the time and ability for. I am a good programmer, but like I said I don't have experience with c++. Still, for the time being, if I can help the project in any way or you have questions, tell me. :smiley:

Oh, and @harsulinsghost drew the mad scientist and allowed me to use it in this post. :smile:

**Comment by numberZero** [comment]
1. That’s enormous work. Incremental adoption is nice, but see (4), for example.
2. What about multithreading?
3. Python-ish typing is attractive for novices, but is too bug-prone. There are way too many examples in mods (as Lua uses almost the same typing system).
4. Exceptions? Cython and C++ are only partially compatible here.
5. Performance is critical, and not only for tight loops, but for some complex classes as well, like MapblockMeshGenerator.

**Comment by zicklag** [comment]
> What about multithreading?

The [documentation](http://docs.cython.org/en/latest/src/userguide/parallelism.html) says that Cython supports native parallelism through OpenMP.

> Python-ish typing is attractive for novices, but is too bug-prone.

I think that that is debatable. There are plenty of programs written in Python and they are not necessarily more bug prone than any other program. Besides you can still use the static C types if necessary.

>  Exceptions? Cython and C++ are only partially compatible here

They way I was thinking that we could implement the conversion would be by individually replacing c++ files with Cython re-implementations that would be compiled to c++ to replace the original c++ file. We could use Cython's [public](http://localhost:5000/src/userguide/external_C_code.html#public-declarations) c type declarations to expose the Cython file with the same declarations as the original c++ file that it replaces ( i.e. they should end up with an identical header file ). I would imagine that any exceptions raised in a public c function in a Cython file would get converted to c exceptions. That should make it completely compatible with the rest of the c++ code base. I'm going to try this out to make sure.

> Performance is critical

I don't believe that very high performance is outside of Cython's reach, after all, it is getting compiled to the same language. If writing a complicated class in Cython really ends up being slower, you can just leave that class written in c++.

An amazing example of high performance Cython is [Spacy](https://spacy.io). Spacy is the fastest natural language processing library in the world and it is written in 100% Cython and Python. I don't think performance will be a problem.

**Comment by zicklag** [comment]
@numberZero, [Here](http://localhost:5000/src/userguide/language_basics.html#error-return-values) is the documentation for having Cython functions raise c++ exceptions. Tell me if you think this is not sufficient.

**Comment by zicklag** [comment]
Personally I don't see any reason that you would ever want to write c++ or c when you could write Cython. I have not found anything that it can't do that c++ or c can do, because, in the end, Cython is compiled to c or c++ and has all of the powers that come with it.

You are right that it is an enormous amount of work. I figure that if you can do it piece by piece then no matter how long it actually takes, eventually you will reap the benefit of having it done.

It is fine if the MineCore team doesn't agree with this proposal and I will understand completely, but I do think that it is a great idea and I will still try it out in my fork, if only to have fun and experiment with the possibility.

**Comment by raymoo** [comment]
OK but you have to do it

**Comment by zicklag** [comment]
@raymoo, Yes, I understand that. In reality it's probably pretty hopeless :sweat_smile:. I was just so intrigued by Cython and how useful it could be that I figured I'd throw it out there and see if anybody else had any thoughts on it.

It probably won't happen, but I'll pursue it until I hit a roadblock or I lose interest. I'll just close this issue for now because it is so far fetched, but I'll still put updates if I make any progress.

**Comment by numberZero** [comment]
@zicklag Thanks for closing this yourself. It’s just unrealistic.
The idea of supporting [CP]ython API doesn’t sound unfeasible, though; *that* may be considered. But see minetest/minetest#4064 for some discussion. I don't agree with many arguments, but that’s still non-straightforward.

> Here is the documentation for having Cython functions raise c++ exceptions.

You linked to localhost. You probably meant http://docs.cython.org/en/latest/src/userguide/language_basics.html#error-return-values. But it is about Python exceptions, not C++ ones. And http://docs.cython.org/en/latest/src/userguide/wrapping_CPlusPlus.html#exceptions tells explicitly that C++ exceptions support is rudimentary. That wouldn’t be a problem for a pure Cython program, but that makes interoperability with current code harder.

> The documentation says that Cython supports native parallelism through OpenMP.

OpenMP? It smells like Fortran. While it may be acceptable for bulk computation but is too restrictive for anything non-straightforward. (Fortunately I never needed to use it, only [MPI](https://en.wikipedia.org/wiki/Message_Passing_Interface) and threads.)

What’s worse, there is GIL. Just read the docs for [parallelism](http://docs.cython.org/en/latest/src/userguide/parallelism.html) and [nogil](http://docs.cython.org/en/latest/src/userguide/external_C_code.html#nogil): “To use this kind of parallelism, the GIL must be released,” and “Code ... must not raise exceptions or manipulate Python objects in any way, and must not call anything that manipulates Python objects without first re-acquiring the GIL.” That is, you can’t speak Python in threads, you can only speak C++ with Python accent.

> An amazing example of high performance Cython is Spacy.

Read [their docs](https://github.com/explosion/spaCy/blob/75d2a05c2938f412f0fae44748374e4de19cc2be/website/usage/_processing-pipelines/_multithreading.jade): “After a long and bitter struggle, the global interpreter lock was freed around spaCy's main parsing loop in v0.100.3.”

Basically, that’s my main concern with Python, Cython, Lua and all that stuff. Only low-level or specially designed language may support parallelism well. In all other cases, you have to do some dirty work, e.g. running independent Lua/JS/whatever instances in separate threads, and maintaining some sort of communication, all that hosted in a low-level environment like C or C++.

**Comment by raymoo** [comment]
OpenMP is good for parallel computation but it is not that good if you want concurrency with multiple threads doing different things, or with execution that doesn't block the execution of another thread (for example, a mapgen thread that can take as long as it wants without blocking the main game loop).

**Comment by zicklag** [comment]
@numberZero @raymoo, Thank you guys for your constructive criticism.My ideas were definitely somewhat naive and based on me doing my best to understand the documentation I read without a true knowledge of low level languages. These are exactly the kind of comments I was hoping to get to help me understand whether or not there were any fundamental limitations to the concept.

Thank you for your detailed explanations of what your concerns are; I now understand a little bit more of how this stuff works. :wink:

**Comment by raymoo** [comment]
I wouldn't have considered my comment constructive

**Comment by zicklag** [comment]
That was bad wording on my part. I didn't take your comment to mean that it was a good idea. The comment was *useful* because it pointed out valid problems with the approach and helped me to understand the shortcomings.

**Comment by zicklag** [comment]
You guys have good points and I plan on dropping the idea completely at this point.
